x/mongo/driver/operation.go

```diff
@@ -1260,6 +1250,35 @@ func (op Operation) addClusterTime(dst []byte, desc description.SelectedServer)
 	// return bsoncore.AppendDocumentElement(dst, "$clusterTime", clusterTime)
 }
 
+// calculateMaxTimeMS calculates the value of the 'maxTimeMS' field to potentially append
+// to the wire message based on the current context's deadline and the 90th percentile RTT
+// if the ctx is a Timeout context. If the context is not a Timeout context, it uses the
+// operation's MaxTimeMS if set. If no MaxTimeMS is set on the operation, and context is
+// not a Timeout context, calculateMaxTimeMS returns 0.
+func (op Operation) calculateMaxTimeMS(ctx context.Context, rtt90 time.Duration) (uint64, error) {
+	if internal.IsTimeoutContext(ctx) {
+		if deadline, ok := ctx.Deadline(); ok {
+			remainingTimeout := time.Until(deadline)
+			// We check that remainingTimeout <= rtt90 because in the rare case that they are equal, we
+			// may accidentally calculate a maxTimeMS value of 0, which would indicate infinite timeout
+			// server-side.
+			if remainingTimeout <= rtt90 {
+				return 0, internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
+					"remaining time %v until context deadline is less than or equal to 90th percentile RTT", time.Until(deadline))
+			}
+			return uint64(remainingTimeout/time.Millisecond - rtt90/time.Millisecond), nil
+		}
+	} else if op.MaxTime != nil {
+		// Users are not allowed to pass a negative value as MaxTime. A value of 0 would indicate
+		// infinite timeout and is allowed.
```

Comment:
I'm sure this is an misunderstanding on my part but why would returning a `op.MaxTime=0` result in an infinite timeout? Wouldn't it just result in a 0 timeout, i.e. `0/x=0, x <> 0`  (the default case of this method)? I think I read in this same PR that

> A maxTimeMS value of 0 only explicitly specifies the default behavior of no timeout server-side.

Or is this just saying that no timeout is the same as an infinite timeout?

Reply:
> Or is this just saying that no timeout is the same as infinite timeout?

Yep! A `maxTimeMS` value of 0 sets an infinite timeout (no timeout) server-side. Similarly in the Go driver, `ClientOptions.SetTimeout(0)` will opt-in to CSOT behavior but will set no context deadline (infinite timeout) on any operation. Semantically, I think "no timeout" and "infinite timeout" are equivalent.

Reply:
Honestly, the term "infinite timeout" is very confusing to me for some reason haha.

Reply:
Understandably, infinite timeout is almost an oxymoron; I'll try to use "no timeout" in documentation and more generally.